// marko/dist/runtime/components/legacy/defineWidget-legacy-browser.js
"use strict";
/* jshint newcap:false */

var BaseState;
var BaseComponent;
var inherit;
var jQuery = require("../jquery");
var ready = require("../ready");

var complain;
var stateToJSONDef = {
  enumerable: false,
  value: function returnSelf() {
    return this;
  }
};
function noop() {}

module.exports = function defineWidget(def, renderer) {
  def = def.Widget || def;

  if (def.___) {
    return def;
  }

  var ComponentClass = function () {};
  var proto;
  var legacyInit;

  if (typeof def === "function") {
    proto = def.prototype;
    legacyInit = def;
  } else if (typeof def === "object") {
    proto = def;
    legacyInit = def.init || noop;
  } else {
    throw TypeError();
  }

  ComponentClass.prototype = proto;

  // We don't use the constructor provided by the user
  // since we don't invoke their constructor until
  // we have had a chance to do our own initialization.
  // Instead, we store their constructor in the "initComponent"
  // property and that method gets called later inside
  // init-components-browser.js
  function Component(id, doc) {
    BaseComponent.call(this, id, doc);
  }

  if (!proto.___) {
    // Inherit from Component if they didn't already
    ComponentClass.prototype = Object.create(BaseComponent.prototype);
    for (var propName in proto) {
      if (proto.hasOwnProperty(propName)) {
        ComponentClass.prototype[propName] = proto[propName];
      }
    }
  }

  // The same prototype will be used by our constructor after
  // we he have set up the prototype chain using the inherit function
  proto = Component.prototype = ComponentClass.prototype;
  proto.aj_ = true;

  proto.constructor = def.constructor = Component;

  Object.defineProperty(proto, "state", {
    get: function () {
      var raw = this.L_ && this.L_._u_;

      if (raw && !raw.toJSON) {
        Object.defineProperty(this.L_._u_, "toJSON", stateToJSONDef);
      }

      return raw;
    },
    set: function (newState) {
      newState = newState || {};
      // eslint-disable-next-line no-constant-condition


      this.setState(newState);
    }
  });

  Object.defineProperty(proto, "__document", {
    get: function () {
      return this.Z_;
      // eslint-disable-next-line no-constant-condition
    }
  });

  Object.defineProperty(proto, "el", {
    get: function () {

      return this.au_;
      // eslint-disable-next-line no-constant-condition
    }
  });

  // get legacy methods

  var legacyOnRender = proto.onRender;
  Object.defineProperty(proto, "onRender", {
    get: noop,
    set: function (v) {
      legacyOnRender = v;
    }
  });

  var legacyOnUpdate = proto.onUpdate;
  Object.defineProperty(proto, "onUpdate", {
    get: function () {
      return modernMountOrUpdate;
    },
    set: function (v) {
      legacyOnUpdate = v;
    }
  });

  var legacyOnDestroy = proto.onDestroy;
  Object.defineProperty(proto, "onDestroy", {
    get: function () {
      return modernOnDestory;
    },
    set: function (v) {
      legacyOnDestroy = v;
    }
  });

  proto.getWidget = proto.getComponent;
  proto.getWidgets = proto.getComponents;
  proto.onMount = modernMountOrUpdate;

  // convert legacy to modern
  var originalUpdate = proto.update;
  proto.update = function () {
    if (this.V_) {

      this.destroy = modernOnDestory;
      // eslint-disable-next-line no-constant-condition

      this.V_ = false;
    }
    this.av_ = true;
    if (this.au_) {
      this.onBeforeUpdate && this.onBeforeUpdate();
    }

    originalUpdate.call(this);
    this.av_ = false;
  };

  function modernMountOrUpdate() {
    var self = this;
    var el = this.getEl("_wbind");
    var prevEl = this.au_;

    if (prevEl !== el) {
      this.au_ = el;

      if (prevEl) {
        this.onBeforeDestroy && this.onBeforeDestroy();
        legacyOnDestroy && legacyOnDestroy.call(this);
        this.removeAllListeners();
      }

      if (el) {
        legacyInit && legacyInit.call(this, this.widgetConfig || {});
        legacyOnRender && legacyOnRender.call(this, { firstRender: true });
        this.on("aw_", function () {
          if (!self.av_) {
            self.onBeforeUpdate && self.onBeforeUpdate();
          }

          self.ax_ = true;
        });

        Object.defineProperty(el, "__widget", {
          configurable: true,
          get: function () {
            return self;
            // eslint-disable-next-line no-constant-condition
          }
        });
      }
    } else if (el) {
      if (prevEl) {
        legacyOnUpdate && legacyOnUpdate.call(this);
      }

      if (this.ax_) {
        legacyOnRender && legacyOnRender.call(this, { firstRender: false });
      }
    }

    this.at_ = this.S_;
    this.S_ = null;
    this.ax_ = false;
  }

  function modernOnDestory() {
    if (this.au_) {
      this.onBeforeDestroy && this.onBeforeDestroy();
      legacyOnDestroy && legacyOnDestroy.call(this);
      this.au_ = null;
    }
  }

  // Set a flag on the constructor function to make it clear this is
  // a component so that we can short-circuit this work later
  Component.___ = true;

  function State() {
    BaseState.apply(this, arguments);
  }
  inherit(State, BaseState);
  proto._d_ = State;

  jQuery.patchComponent(window.$, proto, true /* don't throw error until used if `$` is missing*/
  );

  ready.patchComponent(proto);

  if (!renderer) {
    renderer = ComponentClass.renderer || ComponentClass.prototype.renderer;
    if (renderer) {
      // Legacy support
      var createOut = renderer.createOut;
      if (typeof renderer !== "function") {
        var rendererObject = renderer;
        renderer = function (input, out) {
          var rendererFunc = rendererObject.renderer || rendererObject.render;
          rendererFunc(input, out);
        };
        renderer.createOut = createOut;
      }

      renderer.render = function (input) {
        var out = createOut();
        renderer(input, out);
        return out.end();
      };
    }
  }

  if (renderer) {
    // Add the rendering related methods as statics on the
    // new component constructor function
    Component.renderer = proto._o_ = renderer;
    Component.render = renderer.render;
    Component.renderSync = renderer.renderSync;
  }

  Object.defineProperty(Component, "_isWidget", {
    get: function () {
      return true;
      // eslint-disable-next-line no-constant-condition
    }
  });

  return Component;
};

BaseState = require("../State");
BaseComponent = require("../Component");
inherit = require("raptor-util/inherit");
